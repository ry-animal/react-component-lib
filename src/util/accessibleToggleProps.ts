export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
