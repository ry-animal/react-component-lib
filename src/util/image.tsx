import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
