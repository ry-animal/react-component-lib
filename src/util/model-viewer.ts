/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
