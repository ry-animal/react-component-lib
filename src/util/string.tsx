const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
