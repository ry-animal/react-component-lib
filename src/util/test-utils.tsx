import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
