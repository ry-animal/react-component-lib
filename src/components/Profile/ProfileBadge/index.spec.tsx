import userEvent from "@testing-library/user-event";
import { ProfileBadge } from ".";
import { render, screen } from "../../../util/test-utils";

describe("<ProfileBadge />", () => {
  it("Default rendering matches snapshot", () => {
    expect(
      render(
        <ProfileBadge displayName="smol cat" imgSrc="cat.png" />
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("Default rendering matches snapshot (clickable)", () => {
    expect(
      render(
        <ProfileBadge
          displayName="smol cat"
          imgSrc="cat.png"
          onClick={() => {}}
        />
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("Default rendering matches snapshot (custom sizes)", () => {
    expect(
      render(
        <ProfileBadge
          displayName="smol cat"
          fontSizeRem={0.5}
          iconSizePx={100}
          imgSrc="cat.png"
        />
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("Default rendering matches snapshot (custom color)", () => {
    expect(
      render(
        <ProfileBadge color="red" displayName="smol cat" imgSrc="cat.png" />
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("Default rendering matches snapshot (bold)", () => {
    expect(
      render(
        <ProfileBadge bold displayName="smol cat" imgSrc="cat.png" />
      ).asFragment()
    ).toMatchSnapshot();
  });

  it("with theme swicher", () => {
    render(
      <ProfileBadge
        displayName="smol cat"
        imgSrc="cat.png"
        themeSwitcher={<>Hello</>}
      />
    );
    expect(screen.getByText("Hello")).toBeInTheDocument();
  });

  it("click handler gets called when badge is clicked", () => {
    const onClick = jest.fn();
    render(
      <ProfileBadge displayName="name" imgSrc="cat.png" onClick={onClick} />
    );
    userEvent.click(screen.getByRole("button"));
    expect(onClick).toBeCalled();
  });
});
