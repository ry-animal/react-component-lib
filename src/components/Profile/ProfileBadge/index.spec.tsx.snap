// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileBadge /> Default rendering matches snapshot (bold) 1`] = `
<DocumentFragment>
  .c3 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 36px;
  width: 36px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  border-radius: 50%;
}

.c2 {
  position: relative;
  height: 40px;
  width: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 100%;
  color: #151515;
  gap: 8px;
  padding: auto;
}

.c1:hover {
  background-color: auto;
  border-radius: 4px;
}

.c5 {
  color: #151515;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-family: 'Poppins',sans-serif;
  font-size: 0.875rem;
  font-weight: 600;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
    class="c0"
  >
    <div
      aria-label="View profile"
      class="c1"
    >
      <div
        class="c2"
        size="36"
      >
        <div
          class="c3 c4"
          size="36"
        />
      </div>
      <div
        class="c5"
        title="smol cat"
      >
        smol cat
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProfileBadge /> Default rendering matches snapshot (clickable) 1`] = `
<DocumentFragment>
  .c3 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 36px;
  width: 36px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  border-radius: 50%;
}

.c2 {
  position: relative;
  height: 40px;
  width: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 100%;
  color: #151515;
  gap: 8px;
  padding: 8px 16px 8px 8px;
}

.c1:hover {
  background-color: #F9F9F9;
  border-radius: 4px;
}

.c5 {
  color: #151515;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-family: 'Poppins',sans-serif;
  font-size: 0.875rem;
  font-weight: 500;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
    class="c0"
  >
    <div
      aria-label="View profile"
      class="c1"
      role="button"
    >
      <div
        class="c2"
        size="36"
      >
        <div
          class="c3 c4"
          size="36"
        />
      </div>
      <div
        class="c5"
        title="smol cat"
      >
        smol cat
      </div>
      <svg
        fill="none"
        height="26"
        viewBox="0 0 26 26"
        width="26"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M10.2197 18.2803C9.92678 17.9874 9.92678 17.5126 10.2197 17.2197L14.6893 12.75L10.2197 8.28033C9.92678 7.98744 9.92678 7.51256 10.2197 7.21967C10.5126 6.92678 10.9874 6.92678 11.2803 7.21967L16.2803 12.2197C16.5732 12.5126 16.5732 12.9874 16.2803 13.2803L11.2803 18.2803C10.9874 18.5732 10.5126 18.5732 10.2197 18.2803Z"
          fill="#5F5F5F"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProfileBadge /> Default rendering matches snapshot (custom color) 1`] = `
<DocumentFragment>
  .c3 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 36px;
  width: 36px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  border-radius: 50%;
}

.c2 {
  position: relative;
  height: 40px;
  width: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 100%;
  color: #151515;
  gap: 8px;
  padding: auto;
}

.c1:hover {
  background-color: auto;
  border-radius: 4px;
}

.c5 {
  color: red;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-family: 'Poppins',sans-serif;
  font-size: 0.875rem;
  font-weight: 500;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
    class="c0"
  >
    <div
      aria-label="View profile"
      class="c1"
    >
      <div
        class="c2"
        size="36"
      >
        <div
          class="c3 c4"
          size="36"
        />
      </div>
      <div
        class="c5"
        title="smol cat"
      >
        smol cat
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProfileBadge /> Default rendering matches snapshot (custom sizes) 1`] = `
<DocumentFragment>
  .c3 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 100px;
  width: 100px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  border-radius: 50%;
}

.c2 {
  position: relative;
  height: 104px;
  width: 104px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 100%;
  color: #151515;
  gap: 8px;
  padding: auto;
}

.c1:hover {
  background-color: auto;
  border-radius: 4px;
}

.c5 {
  color: #151515;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-family: 'Poppins',sans-serif;
  font-size: 0.5rem;
  font-weight: 500;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
    class="c0"
  >
    <div
      aria-label="View profile"
      class="c1"
    >
      <div
        class="c2"
        size="100"
      >
        <div
          class="c3 c4"
          size="100"
        />
      </div>
      <div
        class="c5"
        title="smol cat"
      >
        smol cat
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ProfileBadge /> Default rendering matches snapshot 1`] = `
<DocumentFragment>
  .c3 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 36px;
  width: 36px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  border-radius: 50%;
}

.c2 {
  position: relative;
  height: 40px;
  width: 40px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  max-width: 100%;
  color: #151515;
  gap: 8px;
  padding: auto;
}

.c1:hover {
  background-color: auto;
  border-radius: 4px;
}

.c5 {
  color: #151515;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  font-family: 'Poppins',sans-serif;
  font-size: 0.875rem;
  font-weight: 500;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
    class="c0"
  >
    <div
      aria-label="View profile"
      class="c1"
    >
      <div
        class="c2"
        size="36"
      >
        <div
          class="c3 c4"
          size="36"
        />
      </div>
      <div
        class="c5"
        title="smol cat"
      >
        smol cat
      </div>
    </div>
  </div>
</DocumentFragment>
`;
